zetta_utils/layer/volumetric/tensorstore/backend.py

```diff
@@ -0,0 +1,307 @@
+
+IN_MEM_CACHE_NUM_BYTES_PER_TS = 128 * 1024 ** 2
+
+# TODO: Use `assume_metadata` off of the cached info, using `get_info`.
+# Cannot use regular hashkey as the resolutions used need to be tracked
+def _get_ts_at_resolution(
+    path: str, cache_bytes_limit: Optional[int] = None, resolution: Optional[str] = None
+) -> tensorstore.TensorStore:
+    if cache_bytes_limit is None:
+        cache_bytes_limit = IN_MEM_CACHE_NUM_BYTES_PER_TS
+    if (path, resolution) in _ts_cache:
+        return _ts_cache[path, resolution]
+    spec: Dict[str, Any] = {
+        "driver": "neuroglancer_precomputed",
+        "kvstore": path,
+        "context": {"cache_pool": {"total_bytes_limit": cache_bytes_limit}},
+        "recheck_cached_data": "open",
+    }
+    if resolution is not None:
+        spec["scale_metadata"] = {"resolution": ast.literal_eval(resolution)}
+    result = tensorstore.open(spec).result()
+    _ts_cache[(path, resolution)] = result
+    if path not in _ts_cached:
+        _ts_cached[path] = set()
+    _ts_cached[path].add(resolution)
+    return result
+
+
+def _clear_ts_cache(path: str) -> None:
+    resolutions = _ts_cached.pop(path, None)
+    if resolutions is not None:
+        for resolution in resolutions:
+            _ts_cache.pop((path, resolution))
+
+
+@attrs.mutable
+class TSBackend(VolumetricBackend):  # pylint: disable=too-few-public-methods
+    """
+    Backend for peforming IO on Neuroglancer datasts using TensorStore library.
+    Read data will be a ``torch.Tensor`` in ``BCXYZ`` dimension order.
+    Write data is expected to be a ``torch.Tensor`` or ``np.ndarray`` in ``BCXYZ``
+    dimension order.
+    :param path: Precomputed path.
+    :param info_spec: Specification for the info file for the layer. If None, the
+        info is assumed to exist.
+    :param on_info_exists: Behavior mode for when both `info_spec` is given and
+        the layer info already exists.
+
+    """
+
+    path: str
+    info_spec: Optional[PrecomputedInfoSpec] = None
+    on_info_exists: InfoExistsModes = "expect_same"
+    cache_bytes_limit: Optional[int] = None
+
+    def __attrs_post_init__(self):
+        if self.info_spec is None:
+            self.info_spec = PrecomputedInfoSpec()
+        overwritten = self.info_spec.update_info(self.path, self.on_info_exists)
+        if overwritten:
+            _clear_ts_cache(self.path)
+
+    @overload
+    @staticmethod
+    def from_precomputed(backend: TSBackend) -> TSBackend:
+        ...
+
+    @overload
+    @staticmethod
+    def from_precomputed(backend: CVBackend) -> TSBackend:
+        ...
+
+    @overload
+    @staticmethod
+    def from_precomputed(backend: VolumetricSetBackend) -> VolumetricSetBackend:
+        ...
+
+    @overload
+    @staticmethod
+    def from_precomputed(backend: VolumetricBackend) -> VolumetricBackend:
+        ...
+
+    @staticmethod
+    def from_precomputed(backend):  # pragma: no cover
+        if check_type(backend, CVBackend):
```

Comment:
`isinstance` instead of `check_type`?

Reply:
Yeah, I don't know why I used check_type on a non-parametrised type....